### PR TITLE
Adding Models and their specs

### PR DESCRIPTION
**Pull request by SavaVuckovic**
Adding Models and their specs

### CODE REVIEW

File: spec/models/comment_spec.rb
Comment (nidalaa) [inline spec/models/comment_spec.rb]: `expect(subject).to be_valid`

File: spec/models/user_spec.rb
Comment (nidalaa) [inline spec/models/user_spec.rb:49]: Why not:
`subject { users(:one) }`

and

`expect(subject).to be_valid`

File: spec/models/notification_spec.rb
Comment (nidalaa) [inline spec/models/notification_spec.rb]: In V1 there was no notification/activity table 🤔 
Just users, posts, comments and friendships. Lets start with it, please:)

**Review by nidalaa — CHANGES_REQUESTED** [review]

**Comment by nidalaa** [comment]
@SavaVuckovic @talihomz specs look good 😍 

tests


As I understand it you are going with TDD here, right?
In that case I am going to wait for tests to get 🍏 before final PR approval :)

File: app/models/notification.rb
Comment (nidalaa) [inline app/models/notification.rb]: You should not need notifications right now 🤔 why do you want to add it?

File: app/models/post.rb
Comment (nidalaa) [inline app/models/post.rb]: space is missing

File: app/models/user.rb
Comment (nidalaa) [inline app/models/user.rb]: space `presence: true`

File: spec/models/user_spec.rb
Comment (nidalaa) [reply spec/models/user_spec.rb:49]: still

**Review by nidalaa — CHANGES_REQUESTED** [review]

File: app/models/user.rb
Comment (nidalaa) [inline app/models/user.rb]:
```
irb(main):007:0> user.friends_requesting
ActiveRecord::StatementInvalid: PG::UndefinedColumn: ERROR:  column users.friend_id does not exist
LINE 1: SELECT  "users".* FROM "users" WHERE "users"."friend_id" = $...
```

You specified User as class_name, so the query is run on users table. Can you fix it?

Comment (nidalaa) [reply app/models/user.rb]: There is another thing which is not clear for me.

FriendsRequest have status. I assume they can be accepted, waiting or rejected. Am I right?

In that case you should be able to get:
- FriendsRequest send by user X and still waiting
- FriendsRequest send by user X and accepted
- FriendsRequest send by user X and rejected
- FriendsRequest send TO user X and still waiting
- FriendsRequest send TO user X and accepted
- FriendsRequest send TO user X and rejected.

Let's take waiting requests as an example as other ones will be analogical.

You can do it in at least 2 ways:

1)

User model:
```
 has_many :friend_requests_waiting, -> { where "status = 1" }, :class_name => 'FriendRequest', :foreign_key => 'requestor_id'
```
then use it as `User.first.friend_requests_waiting`.

2) User model

```
  has_many :friend_requests, :class_name => 'FriendRequest', :foreign_key => 'requestor_id'
```

FriendRequest model:

```
scope :waiting, -> { where status: 1 }
```

and then use it as `User.first.friend_requests.waiting`.

I think that the latter is better as it keeps FriendsRequest logic inside its model:)

File: app/models/user.rb
Comment (nidalaa) [inline app/models/user.rb]: Friends relationship is very problematic. That's why I am going to save you some time:)

You might try to do a lot of tricks with associations, I recommend some reading of what is possible: http://guides.rubyonrails.org/association_basics.html#choosing-between-has-many-through-and-has-and-belongs-to-many.

However!

In this case user can make friend by:
- send request to someone and wait for acceptance
- OR receive request from someone and accept it.

That means that we would need to look for user id in `requestor_id` column but also in `friend_id`. And there is no way to specify foreign key as (column1 OR column2).

What will be the most clear and reasonable solution?
In **FriendRequest** model:

- add scope

```
scope :mutual_for_user, ->(user) {
    where("(requestor_id = :user_id OR friend_id = :user_id)", user_id: user.id)
  }
```

The you can use:

```
user = User.first
FriendRequest.mutual_for_user(user).accepted
```

In that case 🤔 if you would like to display a list of friends, you will have to take requestor and friend object and display the one who is not the current user.

------------------------------

There is another approach to friendships. We might create to rows for each friendship - so lets say we have user X and Y.

- row 1: requestor_id: X, friend_id Y
- row 2: friend_id: Y, requestor_id: X.

In that case: first row is created when request is sent and second when request is accepted.
We will need to remember to create second row every time the request is accepted.

But you will be able to define all `friends` relationship as 

```
has_many :friends, through: :friend_requests
```

-----------

Please give it a though and decide what do you prefer. I think that both approaches are equally good - we will just need to deal with different challenges.

File: spec/models/user_spec.rb
Comment (nidalaa) [inline spec/models/user_spec.rb]: Actually `Associations` spec do not have much value.
You just test if `has_many :friends` was used in code 🤔 

And if you check it in rails c:

```
irb(main):002:0> u.friends
NameError: uninitialized constant User::Friend
	from (irb):2
irb(main):003:0> 
```

If you really want to test associations, it would be better to:

```
 it 'has many posts' do
    expect(subject.posts).not_to raise_error
 end
```

or even more precise:

```
 it 'has many posts' do
  user_post = Post.create(user_id: subject.id)
  expect(subject.reload.posts,first.id).to eq(user_post.id)
 end
```

**Review by nidalaa — CHANGES_REQUESTED** [review]

File: app/models/user.rb
Comment (nidalaa) [reply app/models/user.rb]: @talihomz  @SavaVuckovic some breaking news here! ;)

Actually you are going to need nice `friends` associations sooner or later.
It means that, we should use second approach - with 2 rows.

You can read about advantages here https://explainextended.com/2009/03/07/selecting-friends/. I really love this article.

You can also try this step by step guide https://dankim.io/mutual-friendship-rails/ which extends my suggestion.

File: app/models/user.rb
Comment (talihomz) [reply app/models/user.rb]: Wow, didn't know that we could do this with Rails!

**Pull request merged**
